bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/ConfigUtils.java

```diff
@@ -103,34 +116,48 @@ public static Object getUnionValue(VariableKey key, BUnionType unionType, String
-                    convertAndGetValuesFromString(matchingValues, TypeConverter::stringToByte, value);
-                    break;
-                case TypeTags.INT_TAG:
-                    convertAndGetValuesFromString(matchingValues, TypeConverter::stringToInt, value);
-                    break;
-                case TypeTags.BOOLEAN_TAG:
-                    convertAndGetValuesFromString(matchingValues, TypeConverter::stringToBoolean, value);
-                    break;
-                case TypeTags.FLOAT_TAG:
-                    convertAndGetValuesFromString(matchingValues, TypeConverter::stringToFloat, value);
-                    break;
-                case TypeTags.DECIMAL_TAG:
-                    convertAndGetValuesFromString(matchingValues, TypeConverter::stringToDecimal, value);
-                    break;
-                case TypeTags.STRING_TAG:
-                    convertAndGetValuesFromString(matchingValues, StringUtils::fromString, value);
-                    break;
-                default:
-                    convertAndGetValuesFromString(matchingValues, TypeConverter::stringToXml, value);
-            }
+            addMatchingValues(type, matchingValues, value);
         }
         return matchingValues;
     }
 
+    private static void addMatchingValues(Type type, List<Object> matchingValues, String value) {
+        Type impliedType = TypeUtils.getImpliedType(type);
+        switch (impliedType.getTag()) {
+            case TypeTags.NULL_TAG:
+                matchingValues.add(BNever.getValue());
+                break;
+            case TypeTags.BYTE_TAG:
+                convertAndGetValuesFromString(matchingValues, TypeConverter::stringToByte, value);
+                break;
+            case TypeTags.INT_TAG:
+                convertAndGetValuesFromString(matchingValues, TypeConverter::stringToInt, value);
+                break;
+            case TypeTags.BOOLEAN_TAG:
+                convertAndGetValuesFromString(matchingValues, TypeConverter::stringToBoolean, value);
+                break;
+            case TypeTags.FLOAT_TAG:
+                convertAndGetValuesFromString(matchingValues, TypeConverter::stringToFloat, value);
+                break;
+            case TypeTags.DECIMAL_TAG:
+                convertAndGetValuesFromString(matchingValues, TypeConverter::stringToDecimal, value);
+                break;
+            case TypeTags.STRING_TAG:
+                convertAndGetValuesFromString(matchingValues, StringUtils::fromString, value);
+                break;
+            case TypeTags.TYPE_REFERENCED_TYPE_TAG:
+                addMatchingValues(type.getCachedReferredType(), matchingValues, value);
+                break;
+            case TypeTags.FINITE_TYPE_TAG:
+                addMatchingValues(TypeUtils.getType(
+                        ((LinkedHashSet<?>) ((BFiniteType) impliedType).getValueSpace()).getFirst()),
+                        matchingValues, value);
```

Comment:
I don’t think it’s correct to call the `addMatchingValues` function again. What we should do instead is loop through the value space and check if there is a matching value. If a match is found, we should add that value to the list and then return immediately, without checking the other types.